statsmodels/regression/special_linear_model.py

```diff
@@ -0,0 +1,182 @@
+
+    """
+
+    # with unbalanced panel
+
+    k_cat = tuple((groups.max(0) + 1).tolist())
+    xm = np.empty(exog_dense.shape[1:] + k_cat)
+    xm.fill(np.nan)
+    xm[:, groups[:, 0], groups[:, 1]] = exog_dense.T
+    keep = ~np.isnan(xm[0]).ravel()
+    finished = False
+    for it in range(max_iter):
+        for axis in range(1, xm.ndim):
+            group_mean = np.nanmean(xm, axis=axis, keepdims=True)
+            xm -= group_mean
+            if np.ptp(group_mean) < atol:
+                finished = True
+                break
+        if finished:
+            break
+
+    xd = xm.reshape(exog_dense.shape[-1], -1).T[keep]
+    if add_mean:
+        xmean = exog_dense.mean(0)
+        xd += xmean
+        xm += xmean[:, None, None]
+    return xm, xd, it
+
+
+class OLSAbsorb(WLS):
+    """OLS model that absorbs categorical explanatory variables
+
+    see docstring for OLS, the following only has the extra parameters for now
+
+    Parameters
+    ----------
+    exog_absorb : ndarray, 1D or 2D
+        categorical, factor variables that will be absorbed
+    absorb_method : string, 'lu' or 'lmgres'
+        method used in projection for absorbing the factor variables.
+        Currently the options use either sparse LU decomposition or sparse
+        `lmgres`
+
+    Notes
+    -----
+    constant: the current parameterization produces a constant when the mean
+    categorical effect is set to zero
+
+    Warning: currently not all inherited methods for OLS are correct.
+    Parameters and inference are correct and correspond to the full model for
+    OLS that includes all factor variables as dummies with zero mean factor
+    effects.
+    """
+
+    def __init__(self, endog, exog, exog_absorb, absorb_method='lu', **kwds):
+        # TODO: does exog_absorb need to be a keyword for from_formula ?
+        # TODO: missing handling and shape, type check for exog_absorb in super
+        absorb = cat2dummy_sparse(exog_absorb)
+        self.projector = PartialingSparse(absorb, method=absorb_method)
+        super(OLSAbsorb, self).__init__(endog, exog, **kwds)
+        # projection is in whiten
+
+        self.k_absorb = absorb.shape[1]
+        #self.df_resid -= self.k_absorb
+        # inline doesn't work, df_resid is property
+        self.df_resid = self.df_resid - self.k_absorb + 1
+        #check this, why + 1
+        self.df_model = self.df_model + self.k_absorb - 1
+
+
+        self.absorb = absorb   # mainly for checking
+
+    def whiten(self, y):
+        # add the mean back in to get a constant
+        # this does not reproduce the constant if fixed effect use reference encoding
+        # It produces the constant if the mean fixed effect is zero
+        y_mean = y.mean(0)
+        return self.projector.partial_sparse(y)[1] + y_mean
+
+
+    def _get_fixed_effects(self, resid):
+        """temporary: recover fixed effects from regression using residuals
+
+        Warning: This uses dense fixed effects dummies at the moment
```

Comment:
this needs to use SparseOLS instead, but in the sparse case we should already have lu-factor or can use gmres